File: apps/Lodgr.Api/Features/Buildings/Controllers/BuildingsController.cs
using Lodgr.Api.Common.Pagination;
using Microsoft.AspNetCore.Mvc;

namespace Lodgr.Api.Features.Buildings;

[ApiController]
[Route("api/buildings")]
public class BuildingsController(IBuildingService service) : ControllerBase
{
    [HttpGet]
    [ProducesResponseType(typeof(PagedResult<BuildingListItemResponse>), StatusCodes.Status200OK)]
    public async Task<ActionResult<PagedResult<BuildingListItemResponse>>> GetBuildings(
        [FromQuery] GetBuildingsRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.GetPagedAsync(request, cancellationToken);
        return Ok(result);
    }

    [HttpGet("{buildingId:long}")]
    [ProducesResponseType(typeof(BuildingDetailResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<BuildingDetailResponse>> GetBuildingById(
        [FromRoute] long buildingId,
        [FromQuery] bool includeDeleted,
        CancellationToken cancellationToken)
    {
        var result = await service.GetByIdAsync(buildingId, includeDeleted, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpGet("{buildingId:long}/rooms")]
    [ProducesResponseType(typeof(PagedResult<BuildingRoomItemResponse>), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<PagedResult<BuildingRoomItemResponse>>> GetBuildingRooms(
        [FromRoute] long buildingId,
        [FromQuery] GetBuildingRoomsRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.GetPagedRoomsAsync(buildingId, request, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost]
    [ProducesResponseType(typeof(BuildingDetailResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<ActionResult<BuildingDetailResponse>> CreateBuilding(
        [FromBody] CreateBuildingRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.CreateAsync(request, cancellationToken);
        return CreatedAtAction(nameof(GetBuildingById), new { buildingId = result.BuildingId }, result);
    }

    [HttpPut("{buildingId:long}")]
    [ProducesResponseType(typeof(BuildingDetailResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<ActionResult<BuildingDetailResponse>> UpdateBuilding(
        [FromRoute] long buildingId,
        [FromBody] UpdateBuildingRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.UpdateAsync(buildingId, request, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{buildingId:long}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> DeleteBuilding(
        [FromRoute] long buildingId,
        CancellationToken cancellationToken)
    {
        var deleted = await service.SoftDeleteAsync(buildingId, cancellationToken);
        return deleted ? NoContent() : NotFound();
    }
}

File: apps/api/src/main/java/org/lodgr/api/features/buildings/BuildingsController.java
package org.lodgr.api.features.buildings;

import static org.lodgr.api.features.buildings.BuildingDtos.*;

import jakarta.validation.Valid;
import org.lodgr.api.pagination.PagedResult;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@Validated
@RequestMapping("/api/buildings")
public class BuildingsController {
    private final BuildingService buildingService;

    public BuildingsController(BuildingService buildingService) {
        this.buildingService = buildingService;
    }

    @GetMapping
    public ResponseEntity<PagedResult<BuildingListItemResponse>> getBuildings(
            @RequestParam(required = false) Integer pageNumber,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) Long ownerId,
            @RequestParam(required = false, defaultValue = "false") Boolean includeDeleted,
            @RequestParam(required = false) String keyword) {
        GetBuildingsQuery query = new GetBuildingsQuery(pageNumber, pageSize, ownerId, includeDeleted, keyword);
        return ResponseEntity.ok(buildingService.getPaged(query));
    }

    @GetMapping("/{buildingId}")
    public ResponseEntity<BuildingDetailResponse> getBuildingById(
            @PathVariable Long buildingId,
            @RequestParam(required = false, defaultValue = "false") boolean includeDeleted) {
        BuildingDetailResponse result = buildingService.getById(buildingId, includeDeleted);
        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @GetMapping("/{buildingId}/rooms")
    public ResponseEntity<PagedResult<BuildingRoomItemResponse>> getBuildingRooms(
            @PathVariable Long buildingId,
            @RequestParam(required = false) Integer pageNumber,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false, defaultValue = "false") Boolean includeDeleted) {
        GetBuildingRoomsQuery query = new GetBuildingRoomsQuery(pageNumber, pageSize, includeDeleted);
        PagedResult<BuildingRoomItemResponse> result = buildingService.getPagedRooms(buildingId, query);

        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @PostMapping
    public ResponseEntity<BuildingDetailResponse> createBuilding(@Valid @RequestBody CreateBuildingRequest request) {
        BuildingDetailResponse result = buildingService.create(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(result);
    }

    @PutMapping("/{buildingId}")
    public ResponseEntity<BuildingDetailResponse> updateBuilding(
            @PathVariable Long buildingId,
            @Valid @RequestBody UpdateBuildingRequest request) {
        BuildingDetailResponse result = buildingService.update(buildingId, request);
        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @DeleteMapping("/{buildingId}")
    public ResponseEntity<Void> deleteBuilding(@PathVariable Long buildingId) {
        boolean deleted = buildingService.softDelete(buildingId);
        if (!deleted) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.noContent().build();
    }
}
